File: frontend/src/types/delivery-exception.ts
import type { CustomerSummaryResponse } from "./customer";

export type ExceptionType = "VACATION" | "NO_MILK" | "HOLIDAY";

export interface SubscriptionSummaryResponse {
  id: number;
  customer: CustomerSummaryResponse;
  morning_quantity: number;
  evening_quantity: number;
}

export interface DeliveryExceptionCreate {
  subscription_id: number;
  exception_type: string;
  shift?: string | null;
  start_date: string;
  end_date?: string | null;
  reason?: string | null;
}

export interface DeliveryExceptionUpdate {
  exception_type?: string;
  shift?: string | null;
  start_date?: string;
  end_date?: string | null;
  reason?: string | null;
  status?: string;
}

export interface DeliveryExceptionListResponse {
  id: number;
  subscription_id: number;
  customer_id: number;
  customer_code: string;
  customer_name: string;
  route_name: string;
  exception_type: string;
  shift: string | null;
  start_date: string;
  end_date: string | null;
  status: string;
  is_active: boolean;
}

export interface DeliveryExceptionResponse {
  id: number;
  subscription_id: number;
  exception_type: string;
  shift: string | null;
  start_date: string;
  end_date: string | null;
  reason: string | null;
  status: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface DeliveryExceptionDetailResponse {
  id: number;
  subscription: SubscriptionSummaryResponse;
  exception_type: string;
  shift: string | null;
  start_date: string;
  end_date: string | null;
  reason: string | null;
  status: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

File: app/schemas/delivery_exception.py
from datetime import datetime

from pydantic import BaseModel
from pydantic import ConfigDict
from pydantic import Field

from app.constants.shifts import Shift
from app.schemas.customer import CustomerSummaryResponse


class SubscriptionSummaryResponse(BaseModel):

    id: int

    customer: CustomerSummaryResponse

    morning_quantity: int

    evening_quantity: int

    model_config = ConfigDict(
        from_attributes=True
    )


class DeliveryExceptionBase(BaseModel):

    subscription_id: int = Field(
        ...,
        gt=0
    )

    exception_type: str = Field(
        ...,
        min_length=1,
        max_length=20
    )

    shift: Shift | None = Field(
        default=None
    )

    start_date: datetime

    end_date: datetime | None = Field(
        default=None
    )

    reason: str | None = Field(
        default=None,
        max_length=255
    )


class DeliveryExceptionCreate(DeliveryExceptionBase):

    pass


class DeliveryExceptionUpdate(BaseModel):

    exception_type: str | None = Field(
        default=None,
        max_length=20
    )

    shift: Shift | None = Field(
        default=None
    )

    start_date: datetime | None = Field(
        default=None
    )

    end_date: datetime | None = Field(
        default=None
    )

    reason: str | None = Field(
        default=None,
        max_length=255
    )

    status: str | None = Field(
        default=None,
        max_length=20
    )


class DeliveryExceptionResponse(DeliveryExceptionBase):

    id: int

    status: str

    is_active: bool

    created_at: datetime

    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True
    )


class DeliveryExceptionListResponse(BaseModel):

    id: int

    subscription_id: int

    customer_id: int

    customer_code: str

    customer_name: str

    route_name: str

    exception_type: str

    shift: str | None

    start_date: datetime

    end_date: datetime | None

    status: str

    is_active: bool

    model_config = ConfigDict(
        from_attributes=True
    )


class DeliveryExceptionDetailResponse(BaseModel):

    id: int

    subscription: SubscriptionSummaryResponse

    exception_type: str

    shift: str | None

    start_date: datetime

    end_date: datetime | None

    reason: str | None

    status: str

    is_active: bool

    created_at: datetime

    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True
    )